## HEART DISEASE AND STROKE PREVENTION  

#### PROBLEM DEFINITION: 

- To analyse the heart disease dataset and do the prediction using machine learning technique.

#### PROBLEM IDENTIFICATION 

- The dataset is publically available on Kaggle website

This dataset is by provided by the Centers for Medicare & Medicaid Services(CDC),which Inpatient and Outpatient claims, Master Beneficiary Summary Files, and many other files.The indicators of this data has been sourced by Centre of Disease Control(CDC), which provide a comprehensive picture of the public health of various Cardio Vscular Diseases(CVDs) and associated risk factors in the United States. The data are organized by location (national and state) and indicator and is further plotted as trends and stratified by sex and race/ethnicity.Dataset includes over 42640 records and 29 attributes.



##### Leaming Repository: 

https://www.kaggle.com/datasets/themrityunjaypathak/heart-disease-and-stroke-prevention

#### ATTRIBUTE INFORMATION INCLUDED-

- Year

- LocationAbbr- Location abbreviation

- LocationDesc- Location description

- DataSource- Abbreviation of data source

- PriorityArea1- Priority Area (Million Hearts® or None)

- PriorityArea2- Priority Area (ABCS or None)

- PriorityArea3- Priority Area (Healthy People 2020 or None)

- PriorityArea4- Priority Area (AHA 2020 Goals: Cardiovascular Health Metrics or None)

- Category- Category description

- Topic- Topic description

- Indicator-Indicator description

- Data_Value_Type-Data Value Type (mean, rate, percentage)

- Data_Value_Unit-Data Value Unit (%, rate per 100,000, etc.)

- Data_Value-Data value (point estimate)

- Data_Value_Alt-Equal to data value, but formatting is numeric

- Data_Value_Footnote_Symbol-Symbol that would be used to flag footnotes

- Data_Value_Footnote-Footnote description

- Confidence_Limit_Low-95% confidence interval lower bound

- Confidence_Limit_High-95% confidence interval upper bound

- Break_Out_Category-Break out category desncription

- Break_Out-Break out group description

- CategoryId-Category lookup value 

- TopicId-Topic lookup value

- IndicatorID-Indicator lookup value

- Data_Value_TypeID-Data value type lookup value

- BreakOutCategoryId-Break out category lookup value

- BreakOutId-Break out group lookup value

- LocationID-Location lookup value

- GeoLocation

#### TARGET :- Toipic
* Different types of heart condition in the United States and  plot the trends and stratified sex and race/ethnicity
           

#### Variable Types:

   - categorical: 23
        
   - Numerical: 6


#### Tools Used
###### The following tools are going to be used for data analysis :

Pandas, Numpy, Matplotlib, Seaborn, scikit-learn !!

 #### 1.1 Fetching Data Directly From The Kaggle:

In [9]:

!pip install opendatasets

In [10]:
import opendatasets as od


In [11]:
dataset = 'https://www.kaggle.com/datasets/themrityunjaypathak/heart-disease-and-stroke-prevention'

In [12]:
od.download(dataset)

Skipping, found downloaded files in "./heart-disease-and-stroke-prevention" (use force=True to force download)


#### 1.2 Import Other Important Liberaries 

In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os


#### 1.3 Objectives 

- To identify the cardiovasular disease and risk factor in different national and states among United states. 

- To see if there is any useful trends between sex/ethnicity.

- To identify any fundamental issue regarding heart disease.

#### 1.4 Load the Heart Disease and Stroke Prevention Dataset

In [14]:
data_dir = './heart-disease-and-stroke-prevention'

In [15]:
os.listdir(data_dir)

['heart_disease_data.csv']

In [16]:
# Read the csv data
heart_data = pd.read_csv('heart_disease_data.csv')


In [17]:
# Current working dictionary
os.getcwd()

'/Users/parul/Capstone-2 Project'

In [18]:
# Display the data 
heart_data

Year LocationAbbr   LocationDesc DataSource   PriorityArea1  \
0      2006           US  United States   Medicare            None   
1      2005           US  United States   Medicare            None   
2      2007           US  United States   Medicare            None   
3      2008           US  United States   Medicare            None   
4      2004           US  United States   Medicare            None   
...     ...          ...            ...        ...             ...   
42635  2012           CO       Colorado   Medicare  Million Hearts   
42636  2010           MI       Michigan   Medicare  Million Hearts   
42637  2012           ME          Maine   Medicare  Million Hearts   
42638  2013           ID          Idaho   Medicare  Million Hearts   
42639  2012           IA           Iowa   Medicare  Million Hearts   

      PriorityArea2 PriorityArea3 PriorityArea4                 Category  \
0              None          None          None  Cardiovascular Diseases   
1              None          None          None  Cardiovascular Diseases   
2              None          None          None  Cardiovascular Diseases   
3              None          None          None  Cardiovascular Diseases   
4              None          None          None  Cardiovascular Diseases   
...             ...           ...           ...                      ...   
42635          None          None          None  Cardiovascular Diseases   
42636          None          None          None  Cardiovascular Diseases   
42637          None          None          None  Cardiovascular Diseases   
42638          None          None          None  Cardiovascular Diseases   
42639          None          None          None  Cardiovascular Diseases   

                        Topic  ... Break_Out_Category           Break_Out  \
0               Heart Failure  ...               Race               Other   
1               Heart Failure  ...               Race               Other   
2      Coronary Heart Disease  ...                Age                 65+   
3      Coronary Heart Disease  ...             Gender              Female   
4               Heart Failure  ...            Overall             Overall   
...                       ...  ...                ...                 ...   
42635                  Stroke  ...             Gender                Male   
42636                  Stroke  ...             Gender              Female   
42637                  Stroke  ...               Race            Hispanic   
42638                  Stroke  ...               Race  Non-Hispanic White   
42639                  Stroke  ...             Gender                Male   

      CategoryId  TopicId  IndicatorID Data_Value_TypeID BreakOutCategoryId  \
0             C1       T5        MD501             Crude              BOC04   
1             C1       T5        MD501             Crude              BOC04   
2             C1       T4        MD301             Crude              BOC03   
3             C1       T4        MD301             Crude              BOC02   
4             C1       T5        MD501             Crude              BOC01   
...          ...      ...          ...               ...                ...   
42635         C1       T6        MD601             Crude              BOC02   
42636         C1       T6        MD601             Crude              BOC02   
42637         C1       T6        MD601             Crude              BOC04   
42638         C1       T6        MD601             Crude              BOC04   
42639         C1       T6        MD601             Crude              BOC02   

       BreakOutId  LocationID                                GeoLocation  
0           RAC07          59                                        NaN  
1           RAC07          59                                        NaN  
2           AGE06          59                                        NaN  
3           GEN02          59                                        NaN  
4   

In [19]:
# Call info method to see a summary of the data 
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42640 entries, 0 to 42639
Data columns (total 29 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Year                        42640 non-null  int64  
 1   LocationAbbr                42640 non-null  object 
 2   LocationDesc                42640 non-null  object 
 3   DataSource                  42640 non-null  object 
 4   PriorityArea1               42640 non-null  object 
 5   PriorityArea2               42640 non-null  object 
 6   PriorityArea3               42640 non-null  object 
 7   PriorityArea4               42640 non-null  object 
 8   Category                    42640 non-null  object 
 9   Topic                       42640 non-null  object 
 10  Indicator                   42640 non-null  object 
 11  Data_Value_Type             42640 non-null  object 
 12  Data_Value_Unit             42640 non-null  object 
 13  Data_Value                  421

Location AND INDICATOR IS OUR Targated parametrer to detect the risk factors.

In [20]:
# To see top 5 rows in the data
heart_data.head()

Year LocationAbbr   LocationDesc DataSource PriorityArea1 PriorityArea2  \
0  2006           US  United States   Medicare          None          None   
1  2005           US  United States   Medicare          None          None   
2  2007           US  United States   Medicare          None          None   
3  2008           US  United States   Medicare          None          None   
4  2004           US  United States   Medicare          None          None   

  PriorityArea3 PriorityArea4                 Category  \
0          None          None  Cardiovascular Diseases   
1          None          None  Cardiovascular Diseases   
2          None          None  Cardiovascular Diseases   
3          None          None  Cardiovascular Diseases   
4          None          None  Cardiovascular Diseases   

                    Topic  ... Break_Out_Category Break_Out CategoryId  \
0           Heart Failure  ...               Race     Other         C1   
1           Heart Failure  ...               Race     Other         C1   
2  Coronary Heart Disease  ...                Age       65+         C1   
3  Coronary Heart Disease  ...             Gender    Female         C1   
4           Heart Failure  ...            Overall   Overall         C1   

   TopicId  IndicatorID Data_Value_TypeID BreakOutCategoryId  BreakOutId  \
0       T5        MD501             Crude              BOC04       RAC07   
1       T5        MD501             Crude              BOC04       RAC07   
2       T4        MD301             Crude              BOC03       AGE06   
3       T4        MD301             Crude              BOC02       GEN02   
4       T5        MD501             Crude              BOC01       OVR01   

   LocationID GeoLocation  
0          59         NaN  
1          59         NaN  
2          59         NaN  
3          59         NaN  
4          59         NaN  

[5 rows x 29 columns]

This dataset contain both numeric and categorical data with heart desease, categorical(-categorical discreption) is our target variable and contains categorical information.

In [21]:
# TO see the attributes 
heart_data.columns.tolist()

['Year',
 'LocationAbbr',
 'LocationDesc',
 'DataSource',
 'PriorityArea1',
 'PriorityArea2',
 'PriorityArea3',
 'PriorityArea4',
 'Category',
 'Topic',
 'Indicator',
 'Data_Value_Type',
 'Data_Value_Unit',
 'Data_Value',
 'Data_Value_Alt',
 'Data_Value_Footnote_Symbol',
 'Data_Value_Footnote',
 'LowConfidenceLimit',
 'HighConfidenceLimit',
 'Break_Out_Category',
 'Break_Out',
 'CategoryId',
 'TopicId',
 'IndicatorID',
 'Data_Value_TypeID',
 'BreakOutCategoryId',
 'BreakOutId',
 'LocationID',
 'GeoLocation']

We hvae seen the null value in GeoLocation column.

### 1.4 Exploring The Data 

In [22]:
# Check the data properties
heart_data.shape


(42640, 29)

In [23]:
# To check the numeric data
heart_data.describe()

Year    Data_Value  Data_Value_Alt  LowConfidenceLimit  \
count  42640.000000  42111.000000    42640.000000        42111.000000   
mean    2008.500000     14.780896       14.572709           12.661509   
std        2.872315     13.286487       13.333826           11.453615   
min     2004.000000      0.400000       -2.000000            0.000000   
25%     2006.000000      4.500000        4.400000            3.800000   
50%     2008.500000     10.700000       10.400000            8.200000   
75%     2011.000000     21.600000       21.500000           19.600000   
max     2013.000000    333.300000      333.300000           94.200000   

       HighConfidenceLimit    LocationID  
count         42111.000000  42640.000000  
mean             17.074028     29.538462  
std              17.910274     16.064317  
min               1.100000      1.000000  
25%               5.000000     16.750000  
50%              12.700000     29.500000  
75%              23.400000     42.500000  
max             710.500000     59.000000

KEY OBSERVATION FROM ABOVE STATISTICAL SUMMARY -

- The Mean is more than the Median (50% percentile) in Date_value, Date_value_Alt,LowConfidence, HighConfidenceLimit, LocationID, That means our data is leftsided skewned. 

- There is large difference in 75th percentile and max values in columns. This indicates, we have outliers present.

- Standard deviation is less than mean in all the columns which indicates data is more disperse. 


In [24]:
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42640 entries, 0 to 42639
Data columns (total 29 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Year                        42640 non-null  int64  
 1   LocationAbbr                42640 non-null  object 
 2   LocationDesc                42640 non-null  object 
 3   DataSource                  42640 non-null  object 
 4   PriorityArea1               42640 non-null  object 
 5   PriorityArea2               42640 non-null  object 
 6   PriorityArea3               42640 non-null  object 
 7   PriorityArea4               42640 non-null  object 
 8   Category                    42640 non-null  object 
 9   Topic                       42640 non-null  object 
 10  Indicator                   42640 non-null  object 
 11  Data_Value_Type             42640 non-null  object 
 12  Data_Value_Unit             42640 non-null  object 
 13  Data_Value                  421

- There is integer, object and float datatypes of variables.
- the second column give us an idea about how big the dataset and how many null values are there for each field.

In [25]:
# To see unique values for each columns
heart_data.nunique()

Year                            10
LocationAbbr                    52
LocationDesc                    52
DataSource                       1
PriorityArea1                    2
PriorityArea2                    1
PriorityArea3                    2
PriorityArea4                    1
Category                         1
Topic                            6
Indicator                       10
Data_Value_Type                  1
Data_Value_Unit                  2
Data_Value                     813
Data_Value_Alt                 814
Data_Value_Footnote_Symbol       1
Data_Value_Footnote              1
LowConfidenceLimit             680
HighConfidenceLimit           1059
Break_Out_Category               4
Break_Out                        9
CategoryId                       1
TopicId                          6
IndicatorID                     10
Data_Value_TypeID                1
BreakOutCategoryId               4
BreakOutId                       9
LocationID                      52
GeoLocation         

In [26]:
heart_data.value_counts(['LocationDesc'])

LocationDesc  
Alabama           820
Alaska            820
New Hampshire     820
New Jersey        820
New Mexico        820
New York          820
North Carolina    820
North Dakota      820
Ohio              820
Oklahoma          820
Oregon            820
Pennsylvania      820
Rhode Island      820
South Carolina    820
South Dakota      820
Tennessee         820
Texas             820
United States     820
Utah              820
Vermont           820
Virginia          820
Washington        820
Washington, DC    820
West Virginia     820
Wisconsin         820
Nevada            820
Nebraska          820
Montana           820
Illinois          820
Arizona           820
Arkansas          820
California        820
Colorado          820
Connecticut       820
Delaware          820
Florida           820
Georgia           820
Hawaii            820
Idaho             820
Indiana           820
Missouri          820
Iowa              820
Kansas            820
Kentucky          820
Louisiana        

- There are same values for all states 

In [27]:
heart_data.value_counts(['Indicator'])

Indicator                                                                                                                                           
Prevalence of all heart disease hospitalizations among all hospitalizations, US Medicare FFS beneficiaries (65+)                                        4680
Prevalence of cerebrovascular disease hospitalizations among all hospitalizations, US Medicare FFS beneficiaries (65+)                                  4680
Prevalence of coronary heart disease hospitalizations among all hospitalizations, US Medicare FFS beneficiaries (65+)                                   4680
Prevalence of heart attack hospitalizations among all hospitalizations, US Medicare FFS beneficiaries (65+)                                             4680
Prevalence of heart failure hospitalizations among all hospitalizations, US Medicare FFS beneficiaries (65+)                                            4680
Prevalence of major cardiovascular disease hospitalizations among 

- There are Prevalence of cerebrovascular disease hospitalizations among all hospitalizations, US Medicare FFS beneficiaries (65+) is 4680
- The rate of hospitalisation with heart failure are same (3640) among adult ages from 65 to 85 and older.

In [28]:
heart_data.value_counts(['Year'])

Year
2004    4264
2005    4264
2006    4264
2007    4264
2008    4264
2009    4264
2010    4264
2011    4264
2012    4264
2013    4264
dtype: int64

- There are same count of cases from 2004 to 2013 that is 4264

In [29]:
heart_data.value_counts(['Topic'])

Topic                                     
Heart Failure                                 19240
Acute Myocardial Infarction (Heart Attack)     4680
Coronary Heart Disease                         4680
Diseases of the Heart (Heart Disease)          4680
Major Cardiovascular Disease                   4680
Stroke                                         4680
dtype: int64

- There are 6 Topics of heart disease which further lead to heart failure. The total heart failure cases are: 19240 

In [30]:
heart_data.value_counts(['Break_Out_Category'])

Break_Out_Category
Race                  20800
Gender                10400
Age                    6240
Overall                5200
dtype: int64

In [31]:
heart_data.value_counts(['Break_Out'])

Break_Out         
Female                5200
Hispanic              5200
Male                  5200
Non-Hispanic Black    5200
Non-Hispanic White    5200
Other                 5200
Overall               5200
65+                   3120
75+                   3120
dtype: int64

In [32]:
# To check the missing values in dataset
heart_data.isnull().sum()

Year                              0
LocationAbbr                      0
LocationDesc                      0
DataSource                        0
PriorityArea1                     0
PriorityArea2                     0
PriorityArea3                     0
PriorityArea4                     0
Category                          0
Topic                             0
Indicator                         0
Data_Value_Type                   0
Data_Value_Unit                   0
Data_Value                      529
Data_Value_Alt                    0
Data_Value_Footnote_Symbol    42111
Data_Value_Footnote           42111
LowConfidenceLimit              529
HighConfidenceLimit             529
Break_Out_Category                0
Break_Out                         0
CategoryId                        0
TopicId                           0
IndicatorID                       0
Data_Value_TypeID                 0
BreakOutCategoryId                0
BreakOutId                        0
LocationID                  

In [33]:
heart_data.isnull().sum().sum()

86629

In [34]:
#Order them (increasing or decreasing) using sort_values
#Call `pd.concat` to present these in a single table (DataFrame) with the helpful column names 'count' and '%'Count (using `.sum()`) the number of missing values (`.isnull()`) in each column of 
#ski_data as well as the percentages (using `.mean()` instead of `.sum()`).

missingdata = pd.concat([heart_data.isnull().sum(), 100 * heart_data.isnull().mean()], axis=1)

missingdata.columns=['count', '%' ]
missingdata.sort_values(by= 'count',ascending = True)  # sort the value in 


count          %
Year                            0   0.000000
BreakOutId                      0   0.000000
BreakOutCategoryId              0   0.000000
Data_Value_TypeID               0   0.000000
IndicatorID                     0   0.000000
TopicId                         0   0.000000
CategoryId                      0   0.000000
Break_Out                       0   0.000000
Break_Out_Category              0   0.000000
LocationID                      0   0.000000
Data_Value_Unit                 0   0.000000
Data_Value_Alt                  0   0.000000
Indicator                       0   0.000000
LocationAbbr                    0   0.000000
LocationDesc                    0   0.000000
DataSource                      0   0.000000
PriorityArea1                   0   0.000000
PriorityArea2                   0   0.000000
PriorityArea3                   0   0.000000
Data_Value_Type                 0   0.000000
PriorityArea4                   0   0.000000
Topic                           0   0.000000
Category                        0   0.000000
HighConfidenceLimit           529   1.240619
LowConfidenceLimit            529   1.240619
Data_Value                    529   1.240619
GeoLocation                   820   1.923077
Data_Value_Footnote         42111  98.759381
Data_Value_Footnote_Symbol  42111  98.759381

- We have total 86,629 missing values in our dataset.
                  
 - There is missing values for Data_Value_Footnote, Data_Value_Footnote_Symbol is 42,111 for each.The mean of each is    98.76%
 
 - There is missing value for Data_vlaues, LowConfidenceLimit, HighConfidenceLimit is 529 for each.The mean of each is     1.2%.
 
 - Missing value for GeoLocation is 820 with mean is 1.9%.

In [35]:
heart_data.drop(['Data_Value_Footnote'], axis=1,inplace = True)

In [36]:
heart_data.drop(['Data_Value_Footnote_Symbol'], axis=1,inplace = True)

In [37]:
heart_data.drop(['Year'], axis=1,inplace = True)

In [38]:
heart_data.drop(['LocationID'], axis=1,inplace = True)

In [39]:
heart_data.shape

(42640, 25)

#### - 1.5 Categorical Data:

In [40]:
heart_data.select_dtypes(object)

LocationAbbr   LocationDesc DataSource   PriorityArea1 PriorityArea2  \
0               US  United States   Medicare            None          None   
1               US  United States   Medicare            None          None   
2               US  United States   Medicare            None          None   
3               US  United States   Medicare            None          None   
4               US  United States   Medicare            None          None   
...            ...            ...        ...             ...           ...   
42635           CO       Colorado   Medicare  Million Hearts          None   
42636           MI       Michigan   Medicare  Million Hearts          None   
42637           ME          Maine   Medicare  Million Hearts          None   
42638           ID          Idaho   Medicare  Million Hearts          None   
42639           IA           Iowa   Medicare  Million Hearts          None   

      PriorityArea3 PriorityArea4                 Category  \
0              None          None  Cardiovascular Diseases   
1              None          None  Cardiovascular Diseases   
2              None          None  Cardiovascular Diseases   
3              None          None  Cardiovascular Diseases   
4              None          None  Cardiovascular Diseases   
...             ...           ...                      ...   
42635          None          None  Cardiovascular Diseases   
42636          None          None  Cardiovascular Diseases   
42637          None          None  Cardiovascular Diseases   
42638          None          None  Cardiovascular Diseases   
42639          None          None  Cardiovascular Diseases   

                        Topic  \
0               Heart Failure   
1               Heart Failure   
2      Coronary Heart Disease   
3      Coronary Heart Disease   
4               Heart Failure   
...                       ...   
42635                  Stroke   
42636                  Stroke   
42637                  Stroke   
42638                  Stroke   
42639                  Stroke   

                                               Indicator  ... Data_Value_Unit  \
0      Prevalence of heart failure hospitalizations a...  ...     Percent (%)   
1      Prevalence of heart failure hospitalizations a...  ...     Percent (%)   
2      Prevalence of coronary heart disease hospitali...  ...     Percent (%)   
3      Prevalence of coronary heart disease hospitali...  ...     Percent (%)   
4      Prevalence of heart failure hospitalizations a...  ...     Percent (%)   
...                                                  ...  ...             ...   
42635  Prevalence of cerebrovascular disease hospital...  ...     Percent (%)   
42636  Prevalence of cerebrovascular disease hospital...  ...     Percent (%)   
42637  Prevalence of cerebrovascular disease hospital...  ...     Percent (%)   
42638  Prevalence of cerebrovascular disease hospital...  ...     Percent (%)   
42639  Prevalence of cerebrovascular disease hospital...  ...     Percent (%)   

      Break_Out_Category           Break_Out CategoryId TopicId IndicatorID  \
0                   Race               Other         C1      T5       MD501   
1                   Race               Other         C1      T5       MD501   
2                    Age                 65+         C1      T4       MD301   
3                 Gender              Female         C1      T4       MD301   
4                Overall             Overall         C1      T5       MD501   
...                  ...                 ...        ...     ...         ...   
42635             Gender                Male         C1      T6       MD601   
42636             Gender              Female         C1      T6       MD601   
42637               Race            Hispanic         C1      T6       MD601   
42638               Race  Non-Hispanic White         C1      T6       MD601   
42639             Gender                Male         C1      T6       MD601   

     

In [41]:
#Checking null values
heart_data.isna().sum().sum()

2407

We have 21 columns left in Categorical data.

This data has no NaN values, no dublicate values, we left with 42640 rows × 25 columns.

          ~~~~~~~~~    "Now dataset have all clean data"    ~~~~~~~~~
          
          
          